#### Intro to Visualization with Python Part 1

In [1]:
## Import libraries 
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

path_data = r'C:\Users\PC\Desktop\CareerFoundry Data Analytics\08-2025 Instacart Basket Analysis'

In [2]:
## Import data 
customers = pd.read_csv(os.path.join(path_data, '02 Data', 'Original Data', 'customers.csv'))

## Wrangling

In [3]:
customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [4]:
customers.shape

(206209, 10)

In [5]:
# Rename the columns
customers = customers.rename(columns={ 
    'Surnam' : 'Last Name', 
    'STATE' : 'State', 
    'date_joined' : 'Date Joined', 
    'n_dependants' : 'Number of Dependents', 
    'fam_status' : 'Family Status', 
    'income' : 'Income'})

In [6]:
customers.head()

user_id First Name Last Name  Gender       State  Age Date Joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Number of Dependents Family Status  Income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

In [7]:
customers.describe()

user_id            Age  Number of Dependents         Income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

In [8]:
# Check for missing values
missing_values = customers.isnull().sum()

In [9]:
# Check for duplicates
duplicates = customers.duplicated()
num_duplicates = duplicates.sum()

In [10]:
num_duplicates

np.int64(0)

There are on duplicates.

In [11]:
missing_values

user_id                     0
First Name              11259
Last Name                   0
Gender                      0
State                       0
Age                         0
Date Joined                 0
Number of Dependents        0
Family Status               0
Income                      0
dtype: int64

The First Name column contains 11259 missing values. However, this does not pose an issue for subsequent analyses, as this field is not relevant to the analytical objectives.

In [12]:
# Check the data types
customers.dtypes

user_id                  int64
First Name              object
Last Name               object
Gender                  object
State                   object
Age                      int64
Date Joined             object
Number of Dependents     int64
Family Status           object
Income                   int64
dtype: object

## Combining

In [13]:
# Import the main dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path_data, '02 Data', 'Prepared Data', 'ords_prods_merged_and_updated.pkl'))

In [14]:
# Combine the dataframes
combined_df = pd.merge(ords_prods_merge, customers, on='user_id', how='inner')

In [15]:
# Check the new df
combined_df.shape

(32404859, 31)

In [16]:
combined_df.head()

C:\Users\PC\anaconda3\Lib\site-packages\pandas\io\formats\format.py:1458: RuntimeWarning: overflow encountered in cast
  has_large_values = (abs_vals > 1e6).any()
C:\Users\PC\anaconda3\Lib\site-packages\pandas\io\formats\format.py:1458: RuntimeWarning: overflow encountered in cast
  has_large_values = (abs_vals > 1e6).any()


order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  ...   order_frequency_flag  \
0                                     Soda  ...  Non-frequent customer   
1  Organic Unsweetened Vanilla Almond Milk  ...  Non-frequent customer   
2                      Original Beef Jerky  ...  Non-frequent customer   
3               Aged White Cheddar Popcorn  ...  Non-frequent customer   
4         XL Pick-A-Size Paper Towel Rolls  ...  Non-frequent customer   

   First Name  Last Name  Gender    State Age  Date Joined  \
0       Linda     Nguyen  Female  Alabama  31    2/17/2019   
1       Linda     Nguyen  Female  Alabama  31    2/17/2019   
2       Linda     Nguyen  Female  Alabama  31    2/17/2019   
3       Linda     Nguyen  Female  Alabama  31    2/17/2019   
4       Linda     Nguyen  Female  Alabama  31    2/17/2019   

  Number of Dependents  Family Status Income  
0                    3        married  40423  
1                    3        married  40423  
2                    3        married  40423  
3                    3        married  40423  
4                    3        married  40423  

[5 rows x 31 columns]

## Export the dataframe in a pkl format

In [17]:
combined_df.to_pickle(os.path.join(path_data, '02 Data','Prepared Data', 'ords_prods_customers.pkl'))